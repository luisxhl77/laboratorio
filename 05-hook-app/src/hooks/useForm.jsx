import { useState } from "react";

export const useForm = (initialForm = {}) => {
    const [formState, setFormState] =  useState(initialForm);

    const onInputChange = ({ target }) => {
        const { name, value } = target;
        setFormState({
            ...formState,
            [name]: value
        })
    }

    const onResetForm = () => {
        setFormState(initialForm);
    }

    return {
        ...formState, // <---- se desestructura desde aqui para no tener que hacerlo en el formulario
        formState,
        onInputChange,
        onResetForm,
    }
}

/*

*/