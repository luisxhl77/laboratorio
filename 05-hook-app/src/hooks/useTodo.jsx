import { useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const initialState = []

const init = () => {
    return JSON.parse( localStorage.getItem( 'todos' )) || [];
}

export const useTodo = () => {
    const [ todos, dispatchTodo ] = useReducer( todoReducer, initialState, init);

    // implementacion de localStore
    useEffect(() => {
      localStorage.setItem('todos', JSON.stringify(todos) || [] );
    }, [todos])
    
    const handleNewTodo = ( todo ) =>{
        alert(`se agrego ${todo.description}`);
        const action = {
            type: '[TODO] Add Todo',
            payload: todo
        }
        dispatchTodo(action)
    }

    const handleDeleteTodo = ( id ) =>{
        alert(`se se elimino el id: ${id}`);
        const action = {
            type: '[TODO] Remove Todo',
            payload: id
        }
        dispatchTodo(action)
    }

    const handleToggleTodo = ( id ) => { 
        console.log(id)
        const action = {
            type: '[TODO] Toggle Todo',
            payload: id
        }
        dispatchTodo(action)
    }

    return {
        todos: todos,
        handleNewTodo: handleNewTodo,
        handleDeleteTodo: handleDeleteTodo,
        handleToggleTodo: handleToggleTodo,
        todosCount: todos.length,
        pendingTodosCount: todos.filter(todo => !todo.done).length,
    }
}