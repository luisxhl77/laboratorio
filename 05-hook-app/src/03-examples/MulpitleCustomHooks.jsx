import { LoadingQuote } from "../componenets/LoadingQuote";
import { Quote } from "../componenets/Quote";
import { useFetch, useCounter } from "../hooks";

export const MulpitleCustomHooks = () => {

    const {counter,increment,decrement,reset,} = useCounter(1)

    const { data, isLoading, hasError} = useFetch(`https://breakingbadapi.com/api/quotes/${counter}`);
    
    const { author, quote } = !!data && data[0]; //si la data tiene un valor entonces toma los datos de la posicion 0
   
   
    return (
        <>
            <h1>BreakingBad Qoutes</h1>
            <hr />

            {
                (isLoading) ? (
                    <LoadingQuote/>  // mostrar pesta√±a de carga
                ) : (
                    <Quote quote={quote} author={author}/> //mostrar busqueda
                )
            }


            <button className="btn btn-primary" onClick={() =>increment(1)} disabled={isLoading}>next quote</button>
            <button className="btn btn-primary" onClick={() =>decrement(1)} disabled={isLoading}>back quote</button>
            <button className="btn btn-primary" onClick={reset} disabled={isLoading}>reset quote</button>

        </>
    )
}
