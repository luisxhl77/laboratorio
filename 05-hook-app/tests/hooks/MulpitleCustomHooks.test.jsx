import { fireEvent, render, screen } from "@testing-library/react";
import { MulpitleCustomHooks } from "../../src/03-examples/MulpitleCustomHooks";
import { useCounter } from "../../src/hooks/useCounter";
import { useFetch } from "../../src/hooks/useFetch";

jest.mock('../../src/hooks/useCounter');
jest.mock('../../src/hooks/useFetch');

describe('Pruebas de <MultipleCustomHooks/> ', () => {

    const mockIncrement = jest.fn();
    useCounter.mockReturnValue({
        counter: 1,
        increment: mockIncrement
    })

    beforeEach( () =>{
        jest.clearAllMocks();
    });

    test('debe de hacer match con el sanpshot', () => {

        useFetch.mockReturnValue({
            data: null,
            isLoading: true,
            hasError: null,
        })

        const { container} = render( <MulpitleCustomHooks /> );
        expect( container).toMatchSnapshot();
        
        const nextButton = screen.getByRole('button',{name: 'next quote'});  //forma de validar si un elemento esta desabilidato
        expect(nextButton.disabled).toBeTruthy();  //forma de validar si un elemento esta desabilidato
    
    })
    
    test('debe de mostrar un Quote', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'fernando',
                quote: 'Hola mundo',
            }],
            isLoading: false,
            hasError: null,
        })

        render( <MulpitleCustomHooks /> );

        expect( screen.getByText('Hola mundo')).toBeTruthy();
        expect( screen.getByText('fernando')).toBeTruthy();

        const nextButton = screen.getByRole('button',{name: 'next quote'});  //forma de validar si un elemento esta desabilidato
        expect(nextButton.disabled).toBeFalsy ();  //forma de validar si un elemento esta desabilidato
         
    })
    
    test('de de llamar lafuncion de incrementar', () => {

        useFetch.mockReturnValue({
            data: [{
                author: 'fernando',
                quote: 'Hola mundo',
            }],
            isLoading: false,
            hasError: null,
        })



        render( <MulpitleCustomHooks /> );

        const nextButton = screen.getByRole('button',{name: 'next quote'});  //forma de validar si un elemento esta desabilidato
        fireEvent.click(nextButton);

        expect( mockIncrement ).toHaveBeenCalled()
    })
})