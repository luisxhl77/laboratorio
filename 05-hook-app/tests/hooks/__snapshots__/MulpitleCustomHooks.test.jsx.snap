// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <MultipleCustomHooks/>  debe de hacer match con el sanpshot 1`] = `
<div>
  <h1>
    BreakingBad Qoutes
  </h1>
  <hr />
  <div
    class="alert alert-info text-center"
  >
    Loading....
  </div>
  <button
    class="btn btn-primary"
    disabled=""
  >
    next quote
  </button>
  <button
    class="btn btn-primary"
    disabled=""
  >
    back quote
  </button>
  <button
    class="btn btn-primary"
    disabled=""
  >
    reset quote
  </button>
</div>
`;
